File: harness/assumptions.py
"""Inputs a KiCad project does not contain, supplied by hand or not at all.

A schematic states what is connected to what and a layout states where the
copper is. Neither states how much current a rail carries, how hot the room is,
or what the dielectric between the layers does - and those are the inputs three
whole domains rest on. Thermal needs a dissipation and an ambient; power
integrity needs a current; signal integrity needs a stackup.

The architecture's rule is that a missing input means the domain is skipped, not
guessed. This is where the inputs come from when they do exist: a short file
beside the board, written by whoever knows the answers.

    boards/stm32-good.assumptions.yml

    ambient_c: 25
    max_junction_c: 125
    rails:
      "+3.3V": 3.3
      "/IN": 12.0
    dissipation_w:
      S1: 1.7
    copper_weight_oz: 1
    stackup:
      layers: 2
      dielectric_mm: 1.5
    high_speed_nets: ["/USB_D+", "/USB_D-"]

Everything in it is optional and nothing is defaulted. An absent `ambient_c`
means the thermal check is skipped and coverage says why; it does not mean 25 C.
Defaulting is the same failure as estimating, one layer further down, and harder
to see because it looks like configuration rather than invention.

No YAML dependency: the format is a small subset - scalars, one level of
mapping, and inline lists - parsed here rather than adding a package for forty
lines of input.
"""

from __future__ import annotations

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent

#: What each gated domain needs before it may run. Stated once, so the gate and
#: the coverage message cannot disagree about it.
REQUIREMENTS = {
    "thermal": ("ambient_c", "dissipation_w"),
    "signal_integrity": ("stackup", "high_speed_nets"),
    "power_integrity": ("stackup", "dissipation_w"),
}


def _scalar(text: str):
    text = text.strip().strip('"').strip("'")
    if text.startswith("[") and text.endswith("]"):
        return [
            part.strip().strip('"').strip("'")
            for part in text[1:-1].split(",")
            if part.strip()
        ]
    if re.fullmatch(r"-?\d+", text):
        return int(text)
    if re.fullmatch(r"-?\d*\.\d+", text):
        return float(text)
    return text


def parse(text: str) -> dict:
    """A small YAML subset: `key: value`, and one level of indented mapping."""
    out: dict = {}
    current: dict | None = None
    for raw in text.split("\n"):
        line = raw.split("#", 1)[0].rstrip()
        if not line.strip():
            continue
        indented = line[:1].isspace()
        if ":" not in line:
            continue
        key, _, value = line.strip().partition(":")
        key = key.strip().strip('"').strip("'")
        value = value.strip()
        if indented and current is not None:
            current[key] = _scalar(value)
        elif value:
            out[key] = _scalar(value)
            current = None
        else:
            current = {}
            out[key] = current
    return out


def load(board_name: str) -> dict | None:
    """The assumptions beside a board, or None when nobody wrote any.

    None and an empty file mean the same thing to everything downstream: the
    gated domains do not run and coverage says which input was missing.
    """
    for suffix in (".assumptions.yml", ".assumptions.yaml", ".assumptions.json"):
        path = ROOT / "boards" / f"{board_name}{suffix}"
        if not path.exists():
            continue
        text = path.read_text(encoding="utf-8")
        return json.loads(text) if suffix.endswith(".json") else parse(text)
    return None


def enabled(inputs: dict | None, research: dict | None = None) -> dict[str, str]:
    """Which gated domains may run, and why each of the rest may not.

    Returns every domain, mapped to "" when it is enabled and to the reason when
    it is not, so the caller can gate and report coverage from one answer.

    Thermal takes a third requirement from a different place. Ambient and a
    dissipation come from the assumptions file, but the junction-to-ambient
    resistance comes from a datasheet, and a thermal reviewer holding two of the
    three has nothing to compute a margin from - it would be reasoning about
    heat in general rather than about this board. A trustworthy figure means one
    the extractor did not mark `low`: those come from multi-column package
    tables where a column was chosen rather than a value read.
    """
    inputs = inputs or {}
    out = {}
    for domain, needs in REQUIREMENTS.items():
        missing = [name for name in needs if not inputs.get(name)]
        out[domain] = "" if not missing else "no " + " and no ".join(
            name.replace("_", " ") for name in missing
        )

    if not out["thermal"]:
        usable = [
            ref
            for ref, record in (research or {}).items()
            if ((record.get("facts") or {}).get("thermal") or {}).get("confidence") != "low"
            and (record.get("facts") or {}).get("thermal")
        ]
        if not usable:
            out["thermal"] = (
                "no trustworthy junction-to-ambient resistance in the researched facts"
            )
    return out

File: site/assumptions.js
/**
 * The inputs a gated reviewer needs, and where the page gets them.
 *
 * The mirror of harness/assumptions.py. Three of the five reviewers cannot say
 * anything true about a board from the board alone. A thermal reviewer needs an
 * ambient and a dissipation before a junction temperature means anything; a
 * signal-integrity reviewer needs a stackup before a trace width is an
 * impedance; a power-integrity reviewer needs both. Absent those, the honest
 * answer is not an estimate with a caveat attached - it is that the domain was
 * not assessed, and it is reported that way.
 *
 * The harness reads them from `boards/<name>.assumptions.yml`. The page has no
 * file to read, so it asks, the same way it asks for datasheet facts: what
 * someone supplies turns a reviewer on, and the graph visibly grows a node.
 * Nothing is defaulted. A field left blank keeps its domain switched off rather
 * than filling in a plausible number, because a number nobody supplied is the
 * one kind of evidence a reviewer cannot tell apart from a measured one.
 */

const KEY = "pcb-eval.assumptions.v1";

/** What each gated domain cannot run without. Mirrors REQUIREMENTS in the Python. */
export const REQUIREMENTS = {
  thermal: ["ambient_c", "dissipation_w"],
  signal_integrity: ["stackup", "high_speed_nets"],
  power_integrity: ["stackup", "dissipation_w"],
};

/** Every input the page can be told, in the order the form shows them. */
export const FIELDS = [
  ["ambient_c", "Ambient", "°C", "25"],
  ["dissipation_w", "Dissipation", "W", "1.2"],
  ["stackup", "Stackup", "", "2-layer, 1.6 mm FR4"],
  ["high_speed_nets", "Fast nets", "", "USB_DP, USB_DM"],
  ["copper_weight_oz", "Copper", "oz", "1"],
];

function read() {
  try {
    return JSON.parse(localStorage.getItem(KEY) || "{}");
  } catch {
    return {};
  }
}

export function inputsFor(boardName) {
  return read()[boardName] || {};
}

/** Store what was supplied, dropping blanks so an empty field stays a gate. */
export function setInputs(boardName, inputs) {
  const all = read();
  const kept = {};
  for (const [name, value] of Object.entries(inputs || {})) {
    const text = typeof value === "string" ? value.trim() : value;
    if (text === "" || text === null || text === undefined) continue;
    kept[name] = text;
  }
  all[boardName] = kept;
  try {
    localStorage.setItem(KEY, JSON.stringify(all));
  } catch {
    /* a private window, or site data blocked. The page still works. */
  }
  return kept;
}

/**
 * Which gated domains may run, and why each of the rest may not.
 *
 * Every domain is returned: "" when it is enabled, the reason when it is not,
 * so one answer both gates the graph and prints as coverage.
 *
 * The Python adds a fourth requirement to thermal - a junction-to-ambient
 * resistance, which is a datasheet number rather than a board one. The page has
 * no way to obtain one yet: it reads PDFs but does not extract parameters from
 * them, and the field that took one by hand is gone. The requirement comes back
 * here when extraction lands.
 */
export function enabled(inputs) {
  const out = {};
  for (const [domain, needs] of Object.entries(REQUIREMENTS)) {
    const missing = needs.filter((name) => !(inputs || {})[name]);
    out[domain] = missing.length
      ? "no " + missing.map((n) => n.replace(/_/g, " ")).join(" and no ")
      : "";
  }
  return out;
}

/** The gates for a board, from everything the page has been told about it. */
export function gatesFor(board) {
  return enabled(inputsFor(board.meta.name));
}
